addons/mail/models/mail_mail.py

```diff
@@ -239,10 +239,26 @@ def process_email_queue(self, ids=None):
         except Exception:
             _logger.exception("Failed processing mail queue")
 
-        # Remove all the <mail.mail> marked as "to delete"
-        self.env['mail.mail'].sudo().search([('to_delete', '=', True)]).unlink()
+        # Remove the <mail.mail> sent marked as "to delete"
+        self.env['mail.mail'].sudo().search(
+            [('to_delete', '=', True), ('id', 'in', ids)]).unlink()
         return res
 
+    @api.autovacuum
+    def _gc_mail_mail(self):
+        """Clean the <mail.mail> marked as "to delete"."""
+        to_delete = self.env['mail.mail'].sudo().search_count([('to_delete', '=', True)])
+
+        _logger.info('Will remove %i <mail.mail>.', to_delete)
+
+        for __ in range(to_delete // 1000 + 1):
```

Comment:
it is better to use `self.env.cr.split_for_in_conditions()`

Reply:
but with this I'll need to fetch all ids isn't it ? (current implementation load only 100 ids)

And the list can be huge...

Reply:
don't need to use the search_count (can be very long)
```python
while True:
    mails = self.env['mail.mail'].sudo().search(
                [('to_delete', '=', True)], order="create_date ASC, id ASC", limit=1000)
    if not mails:
        break
    mails.unlink()
    self.env.cr.commit()
```

Reply:
I preferer to do the search_count because mail.mail can technically be created when looping